modules/cache/context.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package cache
+
+import "context"
+
+type cacheContext struct {
+	ctx  context.Context
+	Data map[any]map[any]any
+}
+
+type cacheContextKey struct{}
+
+func WithCacheContext(ctx context.Context) context.Context {
+	return context.WithValue(ctx, &cacheContextKey{}, &cacheContext{
+		ctx:  ctx,
+		Data: make(map[any]map[any]any),
+	})
+}
+
+func GetContextData(ctx context.Context, tp, key any) any {
+	if ctx == nil {
+		return nil
+	}
+	if c, ok := ctx.Value(&cacheContextKey{}).(*cacheContext); ok {
+		if c.Data[tp] != nil {
+			return c.Data[tp][key]
+		}
+	}
+	return nil
+}
+
+func SetContextData(ctx context.Context, tp, key, value any) {
+	if ctx == nil {
+		return
+	}
+	if c, ok := ctx.Value(&cacheContextKey{}).(*cacheContext); ok {
+		if c.Data[tp] == nil {
+			c.Data[tp] = make(map[any]any)
+		}
+		c.Data[tp][key] = value
+	}
```

Comment:
At least, I can see 2 problems here:

1. It's not ideal to make a new `&cacheContextKey{}` for each call. A global const is needed.
2. There may be concurrency problems.  IIRC there is code like `go foo(ctx)` in routers or services.

Reply:
1 ah, yes. I will fix it.
2 I can there are 2 possible resolutions here. One is to have some check to use contextCache in a go routine, another is to have a lock on the contextCache.

Reply:
This is what I get for reviewing on a phone. I missed these and obsessed about the memory issue!!

Yup,  as per @wxiaoguang  - re: the common cacheContextKey there are examples of this in the other bits of context code.

Re: synchronization - simplest thing would be to stick a sync.RWMutex around any use of c.Data.

Reply:
Done.